[motion-path] Implement offset-position support for circle() / ellipse()

nmoucht opened the pull request:
#### ff4be937a37273cf6ca0a491f28a902406dd78f4
<pre>
[motion-path] Implement offset-position support for circle() / ellipse()
<a href="https://bugs.webkit.org/show_bug.cgi?id=261178">https://bugs.webkit.org/show_bug.cgi?id=261178</a>
rdar://114951375

Reviewed by Simon Fraser.

When the position is not specified for circle or ellipse, use offset-position
instead. Also update normalPositionForOffsetPath to have offset-position: normal
default to the center of the containing block rect for shape path operations.

* LayoutTests/TestExpectations:
* Source/WebCore/css/BasicShapeFunctions.cpp:
(WebCore::basicShapeForValue):
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::MotionPath::computePathForShape):
* Source/WebCore/rendering/MotionPath.h:
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/shapes/Shape.cpp:
(WebCore::Shape::createShape):
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::BasicShapeCircle::floatValueForRadiusInBox const):
(WebCore::BasicShapeCircle::pathForCenterCoordinate const):
(WebCore::BasicShapeCircle::path):
(WebCore::BasicShapeEllipse::pathForCenterCoordinate const):
(WebCore::BasicShapeEllipse::path):
* Source/WebCore/rendering/style/BasicShapes.h:
(WebCore::BasicShapeHasCenterCoordinate::setPositionWasOmitted):
(WebCore::BasicShapeHasCenterCoordinate::positionWasOmitted const):

Canonical link: <a href="https://commits.webkit.org/267810@main">https://commits.webkit.org/267810@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b51d24bb2ba2cdbb018607827991bcfa6c60dbe2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17727 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18057 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18594 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19552 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16583 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17922 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21348 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18207 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18640 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17940 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18239 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15404 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20415 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15472 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16153 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22728 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16483 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16323 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20587 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16890 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14303 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16005 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20369 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2180 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16735 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -5298,6 +5296,11 @@ imported/w3c/web-platform-tests/css/motion/offset-path-shape-xywh-001.html [ Ima
+imported/w3c/web-platform-tests/css/motion/offset-path-shape-circle-005.html [ ImageOnlyFailure ]
+imported/w3c/web-platform-tests/css/motion/offset-path-shape-circle-006.html [ ImageOnlyFailure ]
+imported/w3c/web-platform-tests/css/motion/offset-path-shape-ellipse-005.html
+
```
Can we implement this in the same patch? It should just be 50% 50% if neither `at <position>` or `offset-position` are used.

nmoucht replied:
Yep gonna do that once https://github.com/WebKit/WebKit/pull/16815 lands.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/style/BasicShapes.h:
```diff
@@ -179,7 +189,7 @@ class BasicShapeCircle final : public BasicShape {
     const BasicShapeCenterCoordinate& centerY() const { return m_centerY; }
     const BasicShapeRadius& radius() const { return m_radius; }
-    float floatValueForRadiusInBox(float boxWidth, float boxHeight) const;
+    float floatValueForRadiusInBox(float boxWidth, float boxHeight, float centerX, float centerY) const;
```
Any time you pass a width, height pair ask yourself if passing a Size type would make more sense.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/style/BasicShapes.h:
```diff
@@ -169,7 +169,17 @@ class BasicShapeRadius {
 };
 
-class BasicShapeCircle final : public BasicShape {
+class BasicShapeHasCenterCoordinate : public BasicShape {
```
This is a confusing name for a shape that's a circle or ellipse.

nmoucht replied:
Changed to BasicShapeCircleOrEllipse.

smfr commented on Source/WebCore/rendering/style/BasicShapes.h:
```diff
@@ -203,9 +214,10 @@ class BasicShapeCircle final : public BasicShape {
     BasicShapeCenterCoordinate m_centerX;
     BasicShapeCenterCoordinate m_centerY;
     BasicShapeRadius m_radius;
+    bool centerWasOmitted = false;
```
Do you need this here? It's also on the base class. Also this doesn't have the `m_`.

Can we represent "center omitted" by using std::optional<> for centerX, centerY?

nmoucht replied:
Removed this. I don't think there's a good way to use std::optional here since the code does the conversion from nullptr to the default values for the center of the circle and ellipse shape in `BasicShapeFunctions::basicShapeForValue`. This is a special case for motion path specifically so I think it's simpler to use a bool here.

nmoucht committed:
[motion-path] Implement offset-position support for circle() / ellipse()
https://bugs.webkit.org/show_bug.cgi?id=261178
rdar://114951375

Reviewed by Simon Fraser.

When the position is not specified for circle or ellipse, use offset-position
instead. Also update normalPositionForOffsetPath to have offset-position: normal
default to the center of the containing block rect for shape path operations.

* LayoutTests/TestExpectations:
* Source/WebCore/css/BasicShapeFunctions.cpp:
(WebCore::basicShapeForValue):
* Source/WebCore/rendering/MotionPath.cpp:
(WebCore::MotionPath::computePathForShape):
* Source/WebCore/rendering/MotionPath.h:
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/shapes/Shape.cpp:
(WebCore::Shape::createShape):
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::BasicShapeCircle::floatValueForRadiusInBox const):
(WebCore::BasicShapeCircle::pathForCenterCoordinate const):
(WebCore::BasicShapeCircle::path):
(WebCore::BasicShapeEllipse::pathForCenterCoordinate const):
(WebCore::BasicShapeEllipse::path):
* Source/WebCore/rendering/style/BasicShapes.h:
(WebCore::BasicShapeHasCenterCoordinate::setPositionWasOmitted):
(WebCore::BasicShapeHasCenterCoordinate::positionWasOmitted const):

Canonical link: https://commits.webkit.org/267810@main
